scio-parquet/src/main/scala/com/spotify/scio/parquet/read/ParquetRead.scala

```diff
@@ -43,8 +58,92 @@ object ParquetRead {
     readSupportFactory: ReadSupportFactory[T],
     conf: SerializableConfiguration,
     projectionFn: SerializableFunction[T, R]
-  ): ParDo.SingleOutput[ReadableFile, R] = {
+  ): PTransform[PCollection[ReadableFile], PCollection[R]] = {
     val sdf = new ParquetReadFn[T, R](readSupportFactory, conf, projectionFn)
-    ParDo.of(sdf)
+    val tfx: PTransform[PCollection[_ <: ReadableFile], PCollection[R]] = ParDo.of(sdf)
+
+    tfx.asInstanceOf[PTransform[PCollection[ReadableFile], PCollection[R]]]
+  }
+
+  def readTyped[T: ClassTag: Coder: ParquetType](
+    predicate: FilterPredicate = null,
+    conf: Configuration = null
+  ): PTransform[PCollection[ReadableFile], PCollection[T]] = readTyped[T, T](
+    identity,
+    predicate,
+    conf
+  )
+
+  def readTyped[T: ClassTag: Coder: ParquetType, R](
+    projectionFn: T => R,
+    predicate: FilterPredicate,
+    conf: Configuration
+  ): PTransform[PCollection[ReadableFile], PCollection[R]] = {
+    val configuration = ParquetConfiguration.ofNullable(conf)
+    Option(predicate).foreach(p => ParquetInputFormat.setFilterPredicate(configuration, p))
+
+    val cleanedFn = Functions.serializableFn(ClosureCleaner.clean(projectionFn))
+    readFiles(ReadSupportFactory.typed[T], new SerializableConfiguration(configuration), cleanedFn)
```

Comment:
I'm not really sure the `projectionFn` is doing what it is supposed to.
I checked in the `ParquetReadFn` [here](https://github.com/spotify/scio/blob/ddc4a397dea226ce0132bf2bbff76be6d3382be4/scio-parquet/src/main/scala/com/spotify/scio/parquet/read/ParquetReadFn.scala#L240) and it seems to me it is the equivalent of a chained `map` step.

IMHO we should either:
- simplify the API by removing projection and let user do a `map` post read (I am also surprised here we do not have an implicit `Coder[R]` for the output type)
- ask user to create a model `T` which is a projection of the model used when writing the parquet file

Reply:
Re: implicit `Coder[R]` -- that would be required when the ReadFiles transform is actually applied, here: https://github.com/spotify/scio/blob/main/scio-core/src/main/scala/com/spotify/scio/values/SCollection.scala#L1396

will circle back on the `projectionFn`, I want to check something...

Reply:
ok, so for the projection -- the main advantage of offering the projection, as I understand it, is that it happens pre-Coder#encode.

I think that for GenericRecords, this doesn't offer any advantage, _as long as_ the user correctly specifies an implicit Coder using the projected schema -- see [unit test](https://github.com/spotify/scio/pull/4801/files#diff-8cee8349c982f42099f3e288feab6e0fc8f1031973f33942c34b2df23ec70425R181).

But for SpecificRecords, say you have a schema with two non-nullable fields, but you only want to project one of those fields. Without being able to apply a `projectionFn` pre-Coder#encode, serialization will fail due to the absence of a non-nullable field. I just updated the unit test to use `Account` as a test record, which has non-nullable fields; the SpecificRecord test w/ no projectionFn is currently [failing](https://github.com/spotify/scio/pull/4801/files#diff-8cee8349c982f42099f3e288feab6e0fc8f1031973f33942c34b2df23ec70425R256-R273).

So overall, I think that:

a. ProjectionFn support for GenericRecord reads isn't important, and could be removed; however, I'm worried that users will be confused about setting the correct Coder here.
b. ProjectionFn support should be a requirement for SpecificRecord reads if your projection excludes non-nullable, non-default fields.